### PR TITLE
feat: Add `IsDataPartOriginEntityCode` column on code table

### PR DESCRIPTION
**Pull request by AndreasNicolaisen**
## Description

Work Item ID: [AB#27203](https://dev.azure.com/CluedIn-io/c054b4ae-1dab-43c2-af97-3683c744782f/_workitems/edit/27203)

Add `IsDataPartOriginEntityCode` column to code table. This will indicate if the code is the origin code of any data parts, in the entity being export. The column will be added to existing tables, and is therefore nullable.

Also refactor upgrade code, to be in separate classes.

Also adding archiving of custom table types. since this version will not work with old custom table type.

Related: [AB#27194](https://dev.azure.com/CluedIn-io/c054b4ae-1dab-43c2-af97-3683c744782f/_workitems/edit/27194)

## Test approach 
Update reference, observe that `IsDataPartOriginEntityCode` column is present, and filled with nulls.
Export entities, and observe that `IsDataPartOriginEntityCode` is set.

Unit tests are also added/adjusted.

## Release Note 
feat: Add `IsDataPartOriginEntityCode` column to codes table, to indicate if entitycode is the origin code of at least one datapart.

### CODE REVIEW

**Review by Hyldahl — CHANGES_REQUESTED** [review]
Can we add release upgrade notes like we did with the previous release?

File: src/Connector.SqlServer/Utils/Upgrade/UpgradeTo370Utility.cs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+                // If the table exists, we need to check if it was created in an older version of the connector.
+                // The way we do this is to check if it contains all of the columns that we expect the
+                // main table to have in this version.
+                if ((int)tableExistsResult == 1)
+                {
+                    var tableColumnsSelectText = $"""
+                        SELECT columns.name FROM sys.columns columns
+                        INNER JOIN sys.tables tables
+                        ON tables.object_id = columns.object_id AND 
+                           tables.Name = '{mainTableName.LocalName}' AND
```
Comment (zvirja) [inline src/Connector.SqlServer/Utils/Upgrade/UpgradeTo370Utility.cs:66]: should we also check for schema name?

File: src/Connector.SqlServer/Utils/Upgrade/UpgradeTo370Utility.cs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+                        .ExecuteReaderAsync();
+
+                    var existingColumns = new List<string>();
+                    while (await reader.ReadAsync())
+                    {
+                        existingColumns.Add(reader[0].ToString());
+                    }
+
+                    var expectedColumnNames = MainTableDefinition
+                        .GetColumnDefinitions(StreamMode.Sync, Array.Empty<(string, ConnectorPropertyDataType)>())
```
Comment (zvirja) [inline src/Connector.SqlServer/Utils/Upgrade/UpgradeTo370Utility.cs:82]: Is that OK to hardcode sync mode?

File: src/Connector.SqlServer/Utils/TableDefinitions/EntityCodeExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+namespace CluedIn.Connector.SqlServer.Utils.TableDefinitions;
+
+using CluedIn.Core.Data;
+using CluedIn.Core.Reflection;
+
+using System.Reflection;
+
+internal static class EntityCodeExtensions
+{
+    public static object IsOriginEntityCode(this IEntityCode code)
```
Comment (zvirja) [inline src/Connector.SqlServer/Utils/TableDefinitions/EntityCodeExtensions.cs:12]: nit: Name looks a bit weird, as it return either `bool` or `DBNull.Value`. Should it be normalized to DbValue or something and later reflected somehow in the name?

File: src/Connector.SqlServer/Utils/Upgrade/UpgradeCodeTableUtility.cs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+{
+    internal class UpgradeCodeTableUtility
+    {
+        public static async Task AddIsDataPartOriginEntityCodeToTable(IReadOnlyStreamModel streamModel, SqlName schema, SqlTransaction transaction, ILogger logger)
+        {
+            var codeTableName = TableNameUtility.GetCodeTableName(streamModel, schema);
+            var codeTableColumnsSqlQueryText = $"""
+                SELECT [COLUMN_NAME]
+                FROM INFORMATION_SCHEMA.COLUMNS
+                WHERE TABLE_NAME = N'{codeTableName.LocalName}'
```
Comment (zvirja) [inline src/Connector.SqlServer/Utils/Upgrade/UpgradeCodeTableUtility.cs:23]: Review whether schema should also be compared, as you could have same name in different schemas

**Review by zvirja — COMMENTED** [review]
I left a bit of generic comments I spotted. I have a bit limited knowledge, so hope it makes at least some sense.

Some things to consider:
- Make sure that schema is always considered. There could be two tables with the same name in DB and differ only by schema. Make sure all code is ready for that and will not fail.
- I found that some upgrade utilities here and there rename tables in DB. Can't we re-use the existing archiving logic? Feel like we format archived table name a few times and it could lead to inconsistencies with further refactorings

Otherwise see the comments below

**Review by Hyldahl — APPROVED** [review]

**Pull request merged**